### PR TITLE
NMS-9557: Exceptions in Bridge Discovery

### PR DESCRIPTION
**Pull request by rssntn67**
NMS-9557: Exceptions in Bridge Discovery
This is the last work done to fix all the issue related to enlinkd
bridge discovery.

optimization performed and test have been improved

* JIRA: http://issues.opennms.org/browse/NMS-9557

### CODE REVIEW

**Comment by rssntn67** [comment]
Here are some test results:

4361 nodes
2096 are bridge (with an effective bft

Starting from 0 the bridge discovery is accomplished in 50 minutes .
enlinkd finds:

456      bridgebridgelink
52130  bridgemaclink
76406  ipnettomedia entries

**Comment by rssntn67** [comment]
Improved run on the same network to test load topology e rescan .

4361 nodes
2096 are bridge (with an effective bft)


JAVA_HEAP=1GByte

threads="51"
initial_sleep_time="600000"
rescan_interval="7200000"
use-cdp-discovery="false"
use-bridge-discovery="true"
use-lldp-discovery="false"
use-ospf-discovery="false"
use-isis-discovery="false"
max_bft="600"
discovery-bridge-threads="10"

2018-04-28 09:00:04,058 INFO  [Main] o.o.n.e.EnhancedLinkd: init: Loading Bridge Topology.....
2018-04-28 09:00:11,272 INFO  [Main] o.o.n.e.EnhancedLinkd: init: Bridge Topology loaded.

Topology loaded in 7 seconds.

2018-04-28 09:10:12,510 INFO start enlinkd bridge collections
2018-04-28 09:23:39,878 INFO parsed 2371 nodes in less then 14 minutes
2018-04-28 09:43:57,008 INFO  [EnhancedLinkd-Thread-31-of-51] o.o.n.e.DiscoveryBridgeDomains: run: getting nodes with updated bft on broadcast domains. End

All parsed in 33 minutes.

343 bridgebridgelink
39915 bridgemaclink
69184 ipnettomedia entries

The results are very different from the first run.

**Comment by j-white** [comment]
Nice. Looks promising. I'll take a closer look next week.

**Comment by rssntn67** [comment]
And as a proof that connections are lost I cleaned up the database and restarted opennms.

And I found:

380 bridgebridgelink
43534 brodgemaclink

And in 60 minutes there is no justification for this....
So let's work again on debugging. Hoping is easy to understand what happens.

**Comment by RangerRick** [comment]
One question I have is: should this be retargeted from `foundation-2017`?  You have gone way beyond just simple bug fixes to refactoring a large amount of code and writing essentially new features and support in the Bridge Discovery code.  This should really be targeted to `develop` IMHO because it's way too much to put into Meridian 2017 or even Meridian 2018 without more exposure to real world devices beyond your test bed.

**Comment by rssntn67** [comment]
Rebase on foundation-2018...look like a merge is not easy!
Let's do it!

File: core/schema/src/main/liquibase/foundation-2017/changelog.xml [hunk 1]
```diff
@@ -39,6 +39,22 @@
     </rollback>
   </changeSet>
 
+	<!-- NMS-9559 store forwarders -->
```
Comment (j-white) [inline core/schema/src/main/liquibase/foundation-2017/changelog.xml]: These changes should be in the changelog for foundation-2018 (a new one may need to be created)

File: opennms-model/src/main/java/org/opennms/netmgt/model/topology/BridgeForwardingTable.java [hunk 2]
```diff
@@ -0,0 +1,253 @@
+/*******************************************************************************
+ * This file is part of OpenNMS(R).
+ *
+ * Copyright (C) 2014 The OpenNMS Group, Inc.
```
Comment (j-white) [inline opennms-model/src/main/java/org/opennms/netmgt/model/topology/BridgeForwardingTable.java]: New source files should use the current year in the copyright header.

File: core/schema/src/main/liquibase/foundation-2017/changelog.xml (hunk 1 above)
Comment (rssntn67) [reply core/schema/src/main/liquibase/foundation-2017/changelog.xml]: yup I'm going to do the changes

File: opennms-model/src/main/java/org/opennms/netmgt/model/topology/BridgeForwardingTable.java (hunk 2 above)
Comment (rssntn67) [reply opennms-model/src/main/java/org/opennms/netmgt/model/topology/BridgeForwardingTable.java]: ok

**Comment by rssntn67** [comment]
jesse, what about merging?

**Pull request merged**